### PR TITLE
feat: implement context extraction for selected text in AI popover

### PR DESCRIPTION
**Pull request by sedationh**
## Type of Changes

- [ ] ✨ New feature (feat)
- [ ] 🐛 Bug fix (fix)
- [ ] 📝 Documentation change (docs)
- [ ] 💄 UI/style change (style)
- [ ] ♻️ Code refactoring (refactor)
- [ ] ⚡ Performance improvement (perf)
- [ ] ✅ Test related (test)
- [ ] 🔧 Build or dependencies update (build)
- [ ] 🔄 CI/CD related (ci)
- [ ] 🌐 Internationalization (i18n)
- [ ] 🧠 AI model related (ai)
- [ ] 🔄 Revert a previous commit (revert)
- [ ] 📦 Other changes that do not modify src or test files (chore)

## Related Issue

Closes #

## How Has This Been Tested?

- [ ] Added unit tests
- [ ] Verified through manual testing

## Checklist

- [ ] I have tested these changes locally
- [ ] I have updated the documentation accordingly if necessary
- [ ] My code follows the code style of this project
- [ ] My changes do not break existing functionality
- [ ] If my code was generated by AI, I have proofread and improved it as necessary.

## Additional Information

    

---

## Summary by cubic
Add sentence-level context extraction for selected text and stream word explanations into the AI popover. Improves UX with auto-scroll, loading/error states, and blocked scroll bleed.

- **New Features**
  - Extracts before/selection/after context using sentence boundaries (supports English/Chinese); exposed as HighlightData.
  - Saves the selection Range in state and wires it from SelectionToolbar.
  - Streams analysis from the provider model using a level-aware prompt; shows highlighted context, sticky header, and auto-scrolls as text arrives.
  - Upgrades popover shell: wider layout, max-height with internal scroll, prevents background scroll, and exposes scrollToBottom.
  - Adds comprehensive tests for context extraction across edge cases.

- **Dependencies**
  - Added react-markdown for rendering streamed responses.

### CODE REVIEW

File: apps/extension/src/entrypoints/selection.content/selection-toolbar/ai-button.tsx [hunk 1]
```diff
@@ -32,14 +43,140 @@ export function AiButton() {
...
+  const popoverRef = useRef<PopoverWrapperRef>(null)
+
+  const [aiResponse, setAiResponse] = useState<string>('')
+  logger.info('aiResponse', '\n', aiResponse)
+  const [isLoading, setIsLoading] = useState(false)
+  const [error, setError] = useState<string>('')
+
+  const analyzeSelection = useCallback(async (highlightData: any) => {
+    if (!readProviderConfig || !config) {
+      setError('AI配置未找到')
```
Comment (sedationh) [inline apps/extension/src/entrypoints/selection.content/selection-toolbar/ai-button.tsx]: once this ai feature is ready for users, i will get the i18n finalized

File: apps/extension/src/entrypoints/selection.content/utils.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    ? container.parentElement
+    : (container as Element | null)
+
+  if (!root) {
+    return { before: '', selection: '', after: '' }
+  }
+
+  const fullText = root.textContent ?? ''
+  const selection = selectionRange.toString()
+  const startIndex = fullText.indexOf(selection)
```
[the remark this answers is not in the document]
Comment (sedationh) [reply apps/extension/src/entrypoints/selection.content/utils.ts]: make sense

File: apps/extension/src/components/markdown-renderer.tsx [hunk 3]
```diff
@@ -0,0 +1,81 @@
+import { memo } from 'react'
+import ReactMarkdown from 'react-markdown'
+
+interface MarkdownRendererProps {
+  content: string
+  className?: string
+}
+
+export const MarkdownRenderer = memo(({ content, className = '' }: MarkdownRendererProps) => {
```
Comment (ananaBMaster) [inline apps/extension/src/components/markdown-renderer.tsx]: is memo necessary here?

File: apps/extension/src/entrypoints/selection.content/selection-toolbar/ai-button.tsx [hunk 4]
```diff
@@ -32,14 +43,138 @@ export function AiButton() {
...
+      return
+    }
+
+    setIsLoading(true)
+    setError('')
+    setAiResponse('')
+
+    try {
+      const model = await getReadModel(readProviderConfig.name)
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/selection-toolbar/ai-button.tsx]: In the `main`, now we have to use `getReadModelById` to get the model.

File: apps/extension/src/entrypoints/selection.content/selection-toolbar/components/popover-wrapper.tsx [hunk 5]
```diff
@@ -1,6 +1,6 @@
 import { Icon } from '@iconify/react'
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/selection-toolbar/components/popover-wrapper.tsx:1]: move this file to components folder inside selection-toolbar (create a components folder inside

File: apps/extension/src/entrypoints/selection.content/utils.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
+// 句子边界匹配符号
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/utils.ts]: don;t use chinese comments

File: apps/extension/src/entrypoints/selection.content/selection-toolbar/popover-wrapper.tsx [hunk 7]
```diff
@@ -48,15 +64,57 @@ export function PopoverWrapper({ title, icon, children, onClose, isVisible, setI
...
+      const isAtBottom = scrollTop + clientHeight >= scrollHeight - 1
+
+      // 如果向上滚动且已经在顶部，或者向下滚动且已经在底部，阻止事件传播
+      if ((e.deltaY < 0 && isAtTop) || (e.deltaY > 0 && isAtBottom)) {
+        e.preventDefault()
+        e.stopPropagation()
+      }
+    }
+
+    const handleTouchMove = (e: TouchEvent) => {
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/selection-toolbar/popover-wrapper.tsx]: Why is this necessary? Does this mean that users cannot use touchmove in any circumstance?

File: apps/extension/src/entrypoints/selection.content/selection-toolbar/popover-wrapper.tsx [hunk 8]
```diff
@@ -48,15 +64,57 @@ export function PopoverWrapper({ title, icon, children, onClose, isVisible, setI
...
+    const contentElement = contentRef.current
+    if (!contentElement)
+      return
+
+    const handleWheel = (e: WheelEvent) => {
+      const { scrollTop, scrollHeight, clientHeight } = contentElement
+
+      // 检查是否在滚动边界
+      const isAtTop = scrollTop === 0
+      const isAtBottom = scrollTop + clientHeight >= scrollHeight - 1
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/selection-toolbar/popover-wrapper.tsx]: why need `-1` here? `scrollTop + clientHeight >= scrollHeight` won't work?

File: apps/extension/src/entrypoints/selection.content/utils.ts [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+const sentenceBoundary = /[.!?。？！]/g
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/utils.ts]: this may not able to determine boundary for some languages, is there any existing library for this purpose?

File: apps/extension/src/entrypoints/selection.content/selection-toolbar/ai-button.tsx [hunk 10]
```diff
@@ -32,14 +43,138 @@ export function AiButton() {
...
+          <div className="text-sm text-zinc-700 dark:text-zinc-300 bg-zinc-100 dark:bg-zinc-800 p-3 rounded leading-relaxed">
+            {highlightData?.context.before && (
+              <span className="text-zinc-500 dark:text-zinc-400">
+                {highlightData.context.before}
+              </span>
+            )}
+            {highlightData?.context.selection && (
+              <span
+                className="mx-1.5 rounded font-medium px-0.5"
+                style={{ backgroundColor: 'var(--read-frog-primary)' }}
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/selection-toolbar/ai-button.tsx]: why not why tailwind?

File: apps/extension/src/entrypoints/selection.content/selection-toolbar/ai-button.tsx [hunk 11]
```diff
@@ -32,14 +43,138 @@ export function AiButton() {
...
+              <span className="text-zinc-400 dark:text-zinc-500">无选中内容</span>
+            )}
+          </div>
+        </div>
+        <div className="pt-4">
+          {isLoading && !aiResponse && (
+            <div className="flex items-center justify-center py-8">
+              <div className="flex items-center space-x-3 text-slate-500">
+                <div className="flex space-x-1">
+                  <div className="w-2 h-2 bg-blue-500 rounded-full animate-bounce" style={{ animationDelay: '0ms' }}></div>
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/selection-toolbar/ai-button.tsx:142]: why not use tailwind?

File: apps/extension/src/entrypoints/selection.content/selection-toolbar/ai-button.tsx [hunk 12]
```diff
@@ -32,14 +43,138 @@ export function AiButton() {
...
+
+    // 自动触发AI分析
+    analyzeSelection(highlightData)
+  }, [selectionRange, isVisible, analyzeSelection, setHighlightData])
+
   return (
     <PopoverWrapper
+      ref={popoverRef}
       title="AI"
       icon="hugeicons:ai-innovation-02"
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/selection-toolbar/ai-button.tsx:109]: better use `tabler` icon unless you have a string reason to choose another one

File: apps/extension/src/entrypoints/selection.content/utils.ts [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+  const before = fullText.slice(beforeStart, startIndex).trim()
+
+  const afterEnd = afterBoundary === endIndex + 1 ? fullText.length : afterBoundary
+  const after = fullText.slice(endIndex, afterEnd).trim()
+
+  return { before, selection, after }
+}
+
+export interface HighlightData {
+  type: 'highlight'
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/utils.ts]: any reason to put the type "highlight" here? if no, we may not overdesign this

File: apps/extension/src/entrypoints/selection.content/utils.ts [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+  const match = sentenceBoundary.exec(text)
+  return match ? match.index + 1 : text.length
+}
+
+/**
+ * 获取选中文本的上下文句子
+ */
+export function getContext(selectionRange: Range) {
+  const container = selectionRange.commonAncestorContainer
+  const root = container.nodeType === Node.TEXT_NODE
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/utils.ts:7]: the name `root` is misleading here. it sounds like `document` stuff, which is the root of the whole tree. but it's not

File: apps/extension/src/entrypoints/selection.content/utils.ts [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+    ? container.parentElement
+    : (container as Element | null)
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/utils.ts:9]: should not use `as`. you should consider every possible type and handle them

File: apps/extension/src/entrypoints/selection.content/utils.ts [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+
+  if (!root) {
+    return { before: '', selection: '', after: '' }
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/utils.ts:12]: if no root. should return the whole thing as `null`. Then when we processing the type of `Highlight.data` in the future. the IDE will give us the type `{before: ...., after: ....} | null` and which can help future developers to aware this case immediately.

If you return empty string. People are easily ignore the case. and `null` gives all of us a very straightforward meaning

File: apps/extension/src/entrypoints/selection.content/utils.ts [hunk 17]
```diff
@@ -0,0 +1,112 @@
+// 句子边界匹配符号
+const sentenceBoundary = /[.!?。？！]/g
+
+/**
+ * 计算指定节点在根节点中的文本偏移量
+ */
+function getTextOffset(root: Element, node: Node, offset: number): number {
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/utils.ts]: the name `root` is very misleading, use some name like ancestor, container, ancestorElement instread

File: apps/extension/src/entrypoints/selection.content/utils.ts [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+function findBeforeBoundary(text: string, index: number): number {
+  let boundary = 0
+  let match: RegExpExecArray | null
+
+  // 重置正则表达式的状态
+  sentenceBoundary.lastIndex = 0
+
+  // eslint-disable-next-line no-cond-assign
+  while ((match = sentenceBoundary.exec(text)) !== null) {
+    if (match.index < index) {
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/utils.ts]: swap the if condition and get the cleaner code:
```
      if (match.index! >= index) break
      boundary = match.index! + 1
```

File: apps/extension/src/entrypoints/selection.content/utils.ts [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+  const endIndex = getTextOffset(root, selectionRange.endContainer, selectionRange.endOffset)
+
+  if (startIndex === -1 || endIndex === -1) {
+    return { before: '', selection, after: '' }
+  }
+
+  const beforeBoundary = findBeforeBoundary(fullText, startIndex)
+  const afterBoundary = findAfterBoundary(fullText, endIndex)
+
+  const beforeStart = beforeBoundary === startIndex ? 0 : beforeBoundary
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/utils.ts]: this have 2 problems

1. the meaning of `boundary` and `start` is not clear. please use longer name to express what truly the differences between them
2. If there is a case that `This is a sentence. People. Three. Four.` And if the startIndex is the spece in ` Three`. Then `beforeStart` will be 0 and will select the sentence `This is a sentence. People. Three` which is not desirable I believe.

CleanShot 2025-09-19 at 20 48 43@2x

Also please add more test in `ai-button.test.ts` for more cases

File: apps/extension/src/entrypoints/selection.content/utils.ts [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+  const before = fullText.slice(beforeStart, startIndex).trim()
+
+  const afterEnd = afterBoundary === endIndex + 1 ? fullText.length : afterBoundary
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/utils.ts]: please confirm if the logic here is correct for all corner cases. use more test to test it

File: apps/extension/src/entrypoints/selection.content/selection-toolbar/__tests__/ai-button.test.ts [hunk 21]
```diff
@@ -0,0 +1,224 @@
...
+      const range = createRangeForText(testElement, 'This is selected text')
+      const result = getContext(range)
+
+      expect(result.selection).toBe('This is selected text')
+      expect(result.before).toBe('')
+      expect(result.after).toBe('. This is the third sentence.')
+    })
+
+    it('should handle English sentence boundaries', () => {
+      testElement.innerHTML = 'This is the first sentence. This is selected text. This is the third sentence.'
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/selection-toolbar/__tests__/ai-button.test.ts]: same test as the previous one. 
suggest, in the previous test. only test select `text` in `This is selected text`. because we want to make every test aligns with their description. one test for one functionality.

File: apps/extension/src/entrypoints/selection.content/selection-toolbar/ai-button.tsx [hunk 22]
```diff
@@ -32,14 +43,138 @@ export function AiButton() {
...
+        highlightData,
+      )
+
+      const result = await streamText({
+        model,
+        prompt,
+      })
+
+      for await (const delta of result.textStream) {
+        setAiResponse(prev => prev + delta)
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/selection-toolbar/ai-button.tsx]: don't use `setState` in `useEffect`. you call `analyzeSelection` in `useEffect` and which will call `setState` like `setAiResponse`, `setIsLoading` inside.

We should strongly avoid `setState` in `useEffect` since it causes many problem like race condition, unnecessary rerendering.

Instead, use tanstack query to handle all these stuff

File: apps/extension/src/entrypoints/selection.content/selection-toolbar/ai-button.tsx [hunk 23]
```diff
@@ -32,14 +43,138 @@ export function AiButton() {
...
+  const [highlightData, setHighlightData] = useState<HighlightData | null>(null)
+  const popoverRef = useRef<PopoverWrapperRef>(null)
+
+  const [aiResponse, setAiResponse] = useState<string>('')
+  const [isLoading, setIsLoading] = useState(false)
+  const [error, setError] = useState<string>('')
+
+  logger.info('aiResponse', '\n', aiResponse)
+
+  const analyzeSelection = useCallback(async (highlightData: any) => {
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/selection-toolbar/ai-button.tsx]: the useCallback for analyzeSelection is  unnecessary here.

The function is:
1. Only used inside a useEffect
2. The useEffect already has proper dependencies
(selectionRange, isVisible, analyzeSelection,
setHighlightData)
3. The function captures readProviderConfig and config which
are also in the useEffect dependencies

File: apps/extension/src/utils/prompts/word-explain.ts [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+  return `# Identity
+
+You are a professional ${sourceLang} language teacher who provides clear and concise explanations for words and phrases. Your student speaks ${targetLang}. Your student's language level is ${langLevel}.
+
+# Variables
+
+- sourceLang: ${sourceLang}
+- targetLang: ${targetLang}
+- langLevel: ${langLevel}
+- selection: ${highlightData.context.selection}
```
[the remark this answers is not in the document]
Comment (ananaBMaster) [reply apps/extension/src/utils/prompts/word-explain.ts:19]: @sedationh do you think this is reasonable or not?

File: apps/extension/src/utils/prompts/word-explain.ts [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+7. Adjust complexity based on student's language level:
+   - For beginner level: Use simple vocabulary and provide more basic explanations
+   - For intermediate level: Provide balanced explanations with some advanced concepts
+   - For advanced level: Focus on nuanced meanings and sophisticated usage
+
+# Selected Content
+"${highlightData.context.selection}"
+
+# Context
+"${highlightData.context.before}${highlightData.context.selection}${highlightData.context.after}"
```
[the remark this answers is not in the document]
Comment (ananaBMaster) [reply apps/extension/src/utils/prompts/word-explain.ts:43]: This seems to make sense? @sedationh

File: apps/extension/src/entrypoints/selection.content/selection-toolbar/ai-button.tsx [hunk 26]
```diff
@@ -32,14 +43,140 @@
...
+  }, [readProviderConfig, config])
+
+  useEffect(() => {
+    if (!selectionRange || !isVisible) {
+      return
+    }
+
+    const highlightData = createHighlightData(selectionRange)
+    // eslint-disable-next-line no-console
+    console.log('%c seda [ highlightData.context ]-46', 'font-size:13px; background:pink; color:#bf2c9f;', highlightData.context)
```
[the remark this answers is not in the document]
Comment (ananaBMaster) [reply apps/extension/src/entrypoints/selection.content/selection-toolbar/ai-button.tsx]: if you really want to log something, use `logger.log`

[the remark this answers is not in the document]
Comment (ananaBMaster) [reply apps/extension/src/entrypoints/selection.content/selection-toolbar/ai-button.tsx]: @cubic-dev-ai, I'm not talking to you

**Comment by mengxi-ream** [comment]
add pnpm changset

File: apps/extension/src/entrypoints/selection.content/selection-toolbar/ai-button.tsx [hunk 27]
```diff
@@ -32,14 +43,130 @@ export function AiButton() {
...
+    if (!selectionRange || !isVisible) {
+      return null
+    }
+    const data = createHighlightData(selectionRange)
+    logger.info('highlightData.context', '\n', data.context)
+    return data
+  }, [selectionRange, isVisible])
+
+  const {
+    isLoading,
```
Comment (ananaBMaster) [inline apps/extension/src/entrypoints/selection.content/selection-toolbar/ai-button.tsx:62]: use `isPending` here

**Review by ananaBMaster — CHANGES_REQUESTED** [review]
add error handling or comment the ai button component out
